File: language/php/fib.php
<?php
function fib($n)
{
    return $n < 2 ? $n : fib($n-1) + fib($n-2);
}

function fibLinear($n, $prevFib = 0, $fib = 1)
{
    if ($n < 2) {
        return $n;
    }
    while (--$n) {
        [$prevFib, $fib] = [$fib, $fib + $prevFib];
    }
    return $fib;
}

function fibFormula($n)
{
    return round(((5 ** .5 + 1) / 2) ** $n / 5 ** .5);
}

function fibTailRecursive($n, $prevFib = 0, $fib = 1)
{
    return $n === 0 ? $prevFib : fibTailRecursive($n - 1, $fib, $prevFib + $fib);
}

function fibFastDoubling($n)
{
    function fibFastAuxiliary($n)
    {
        if ($n == 0) {
            return [0, 1];
        }
        [$prevFib, $fib] = fibFastAuxiliary(floor($n / 2));
        $a = $prevFib * ($fib * 2 - $prevFib);
        $b = $prevFib * $prevFib + $fib * $fib;
        return $n % 2 == 0 ? [$a, $b] : [$b, $a + $b];
    }
    return fibFastAuxiliary($n)[0];
}

function fibGenerator($n, $prevFib = 0, $fib = 1)
{
    while ($n-- >= 0) {
        yield $prevFib;
        [$prevFib, $fib] = [$fib, $fib + $prevFib];
    }
}
function fibGenerate($n)
{
    foreach (fibGenerator($n) as $fib) echo $fib.PHP_EOL;
}

$input = count($argv) > 1 ? $argv[1] : 29;
printf("Fibonacci %d: %d".PHP_EOL, $input, fibFastDoubling($input));

File: language/ruby/fib.rb
def fib(n)
  n < 2 ? n : fib(n - 1) + fib(n - 2)
end

def fibLinear(n, prevFib = 0, fib = 1)
  return n if n < 2
  while n > 1 do
    prevFib, fib = fib, prevFib + fib
    n -= 1
  end
  fib
end

def fibFormula(n)
  ((((5 ** 0.5 + 1) / 2) ** n / 5 ** 0.5) + 0.5).round
end

def fibTailRecursive(n, prevFib = 0, fib = 1)
  n == 0 ? prevFib : fibTailRecursive(n - 1, fib, prevFib + fib)
end

def fibFastDoubling(n)
  def fibFastAuxiliary(n)
    return [0, 1] if n == 0
    prevFib, fib = fibFastAuxiliary((n / 2).floor)
    a = prevFib * (fib * 2 - prevFib)
    b = prevFib * prevFib + fib * fib
    n % 2 == 0 ? [a, b] : [b, (a + b)]
  end
  fibFastAuxiliary(n)[0]
end

input = ARGV.length > 0 ? ARGV[0].to_i : 29
puts fibFastDoubling(input)
